### PR TITLE
Fix an IRGen bug related to re-use of type metadata

### PR DESCRIPTION
**Pull request by swiftix**
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

---

Before merging this pull request to apple/swift repository:
- [x] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| All supported platforms | @swift-ci Please smoke test and merge |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

**Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| All supported platforms | @swift-ci Please test and merge |
| OS X platform | @swift-ci Please test OS X platform |
| OS X platform | @swift-ci Please benchmark |
| Linux platform | @swift-ci Please test Linux platform |

**Lint Testing**

| Language | Comment |
| --- | --- |
| Python | @swift-ci Please Python lint |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

If Self type in a protocol conformance is a class which may have derived classes, it is not safe to assume that the metadata is exact for this type.

rdar://27301453

### CODE REVIEW

**Comment by swiftix** [comment]
@swift-ci Please test

**Comment by swiftix** [comment]
@rjmccall John, do you mind reviewing this small bugfix? It is in your IRGen code for the dominance-based metadata optimizations.

File: lib/IRGen/GenProto.cpp [hunk 1]
```diff
@@ -511,7 +511,14 @@ void EmitPolymorphicParameters::bindExtraSource(const MetadataSource &source,
       auto selfTy = FnType->getSelfInstanceType();
       CanType argTy = getTypeInContext(selfTy);
       setTypeMetadataName(IGF.IGM, metadata, argTy);
-      IGF.bindLocalTypeDataFromTypeMetadata(argTy, IsExact, metadata);
+      auto *CD = selfTy.getClassOrBoundGenericClass();
+      // If it is a class and it is not final, the metadata is not exact
+      // for argTy, because it may be a metadata for a derived class.
+      // In all other cases, the metadata is exact.
+      // TODO: If it can be proved that the class has no subclasses 
+      // (e.g. in WMO mode), then the metadata can be considered exact.
```
Comment (rjmccall) [inline lib/IRGen/GenProto.cpp]: I would suggest this comment:

The self metadata here corresponds to the conforming type.  For an inheritable conformance, that may be a subclass of the static type, and so the self metadata will be inexact.  Currently, all conformances are inheritable.

**Comment by rjmccall** [comment]
Conceptually LGTM.

File: lib/IRGen/GenProto.cpp (hunk 1 above)
Comment (swiftix) [inline lib/IRGen/GenProto.cpp]: I'll update the comment.

**Comment by swiftix** [comment]
@rjmccall Thanks for reviewing it!

**Comment by swiftix** [comment]
@swift-ci Please test and merge.

**Comment by swiftix** [comment]
@swift-ci Please test and merge

**Pull request merged**